File: clients/scala/src/main/scala/model/Mine.scala
package model

import util.StreamUtil

case class Mine(playerId: Int, position: model.Vec2Double, size: model.Vec2Double, state: model.MineState, timer: Option[Double], triggerRadius: Double, explosionParams: model.ExplosionParams) {
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, playerId)
        position.writeTo(stream)
        size.writeTo(stream)
        state.writeTo(stream)
        timer match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                StreamUtil.writeDouble(stream, value)
            }
        }
        StreamUtil.writeDouble(stream, triggerRadius)
        explosionParams.writeTo(stream)
    }
}
object Mine {
    def readFrom(stream: java.io.InputStream): Mine = Mine(
        StreamUtil.readInt(stream)
        ,
        model.Vec2Double.readFrom(stream)
        ,
        model.Vec2Double.readFrom(stream)
        ,
        model.MineState.readFrom(stream)
        ,
        if (StreamUtil.readBoolean(stream)) Some(
            StreamUtil.readDouble(stream)
        ) else None
        ,
        StreamUtil.readDouble(stream)
        ,
        model.ExplosionParams.readFrom(stream)
        )
}

File: clients/javascript/model/mine.js
const Vec2Double = require('./vec2-double').Vec2Double;
const ExplosionParams = require('./explosion-params').ExplosionParams;
class Mine {
    constructor(playerId, position, size, state, timer, triggerRadius, explosionParams) {
        this.playerId = playerId;
        this.position = position;
        this.size = size;
        this.state = state;
        this.timer = timer;
        this.triggerRadius = triggerRadius;
        this.explosionParams = explosionParams;
    }
    static async readFrom(stream) {
        let playerId;
        playerId = await stream.readInt();
        let position;
        position = await Vec2Double.readFrom(stream);
        let size;
        size = await Vec2Double.readFrom(stream);
        let state;
        state = stream.readInt();
        let timer;
        if (await stream.readBool()) {
            timer = await stream.readDouble();
        } else {
            timer = null;
        }
        let triggerRadius;
        triggerRadius = await stream.readDouble();
        let explosionParams;
        explosionParams = await ExplosionParams.readFrom(stream);
        return new Mine(playerId, position, size, state, timer, triggerRadius, explosionParams);
    }
    async writeTo(stream) {
        let playerId = this.playerId;
        await stream.writeInt(playerId);
        let position = this.position;
        await position.writeTo(stream);
        let size = this.size;
        await size.writeTo(stream);
        let state = this.state;
        await stream.writeInt(state);
        let timer = this.timer;
        if (timer === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await stream.writeDouble(timer);
        }
        let triggerRadius = this.triggerRadius;
        await stream.writeDouble(triggerRadius);
        let explosionParams = this.explosionParams;
        await explosionParams.writeTo(stream);
    }
}
module.exports = { Mine: Mine }
